### ISSUE TITLE
mmdeploy onnx with nms at the end

### ISSUE THREAD

**Issue by shuyuan-wang** [openppl-public/ppq#377]
I exported the yolox_s onnx file with postprocessing using mmdeploy which has TRTBatchedNMS operator at the end, does your onnx-trt conversion support this?

**Comment by ZhangZhiPku** [comment]
如果你的模型中包含自定义算子，你将至少遇到两个问题：
1. PPQ不认你这个算子
2. 推理框架不认你这个算子

为了解决PPQ这边的问题，你可以 from ppq.lib import register_operation_handler，然后注册一个 TRTBatchedNMS 的实现给 ppq。当然这个层往往是最后一层，所以你可以随便写一个空的实现，它并不会影响量化结果。

为了解决推理框架的问题，你也需要给推理框架注册这个算子，就tensorrt而言，你需要为他注册插件。
batchedNMS是一个官方插件，你可以在这里找到它的实现 https://github.com/NVIDIA/TensorRT/tree/main/plugin/batchedNMSPlugin

你需要编译这个东西，并且在调用tensorrt前加载他编译好的.so 或者.dll文件。

**Comment by shuyuan-wang** [comment]
你指的空的实现是指直接return吗？因为mmdeploy出来的output shape是：
![截屏2023-02-08 09 33 20](https://user-images.githubusercontent.com/104115572/217405837-1bb5c1ad-38bb-4545-98ed-1d94aa120195.png)
这个shape不能直在register_operation_handler里面写吧

我导出的onnx是这样的，麻烦帮忙看一下：
链接: https://pan.baidu.com/s/1RrWXRMQFRoSVM-ldYWQgLQ?pwd=jy3y 提取码: jy3y

**Comment by ZhangZhiPku** [comment]
return [torch.zeros(size=[1, 1, 1, 1]), torch.zeros(size=[1, 1, 1, 1])] 就行。
在导出模型之前，执行下面的操作覆写variable的shape：

graph.variables['dets'].shape = ['Batchsize', 'dim_1', 'dim_2']
graph.variables['labels'].shape = ['Batchsize', 'dim_1', 'dim_2']

**Comment by shuyuan-wang** [comment]
现在load_plugin以后可以转engine了，但是在转的时候貌似全部出现了missing scale and zero-point for tensor
![截屏2023-02-08 11 26 49](https://user-images.githubusercontent.com/104115572/217421407-d022e8a0-7a9f-4e60-b884-60b64ce1d032.png)

用的文件是这些；
链接: https://pan.baidu.com/s/1W9PIlucVtg-ivpeMAXfaJg?pwd=xiqr 提取码: xiqr 

最后转的command是：
python ppq/utils/write_qparams_onnx2trt.py --onnx quantized.onnx --qparam_json quant_cfg.json --engine yolox_s_ppq.engine

**Comment by ZhangZhiPku** [comment]
那个倒是没啥问题，我想你的网络已经成功转入int8了。

TensorRT会对所有图中的variable都要一个scale和zp，但是显然不是所有variable都能量化，比如nms的输入和输出变量，因此出现上述一堆scale和zp缺失的情况是符合预期的。

**Comment by shuyuan-wang** [comment]
成功了，感谢
还有一个问题。我在看英伟达的pytorch_quantization的时候发现他对resnet应该是有进一步的optimization：
https://docs.nvidia.com/deeplearning/tensorrt/pytorch-quantization-toolkit/docs/tutorials/quant_resnet50.html#further-optimization

想问一下ppq里面包含了这种优化了吗

**Comment by ZhangZhiPku** [comment]
我速览了一下nvidia文档中的内容，这部分内容关于 conv-add 融合，而你并不需要在 PPQ 中进行更多的操作，这是因为：

1. PPQ 使用 json 传递量化信息，而非采用 qdq 节点的方式，只有在使用 qdq 节点传递量化信息时才会因为错误的插入位置而导致 conv-add 融合被破坏。

**Comment by shuyuan-wang** [comment]
志佬，yolox_l量化出来比fp16还慢，这个正常吗

**Comment by shuyuan-wang** [comment]
> 你可能需要发原始的yolox_l.onnx给我，这个带了qdq节点的模型我已经无法读取到ppq里面了

链接: https://pan.baidu.com/s/17r-rIe8_M8X7zzanSILtQQ?pwd=vsk8 提取码: vsk8

**Comment by ZhangZhiPku** [comment]
在我这里测试，int8比fp16快30%左右。
我的测试环境是 
PPQ 0.6.6
TensorRT 8.5.2.2
nvidia-cublas-cu11       11.10.3.66
nvidia-cuda-nvrtc-cu11   11.7.99
nvidia-cuda-runtime-cu11 11.7.99
nvidia-cudnn-cu11        8.5.0.96

GPU: Nvidia A100
OS: Centos

"""
这是一个可高度定制化的 PPQ 程序入口脚本
与 ProgramEntrance_1 相比，新的 API 允许你控制量化的所有细节。
你可以：
    * 自定义与注册新的量化器
    
    * 干预网络调度逻辑
    
    * 在量化之前修改网络结构
    
    * 自定义优化过程，并控制量化管线
"""

import os

import numpy as np
import torch
import torchvision

import ppq.lib as PFL
from ppq import TargetPlatform, TorchExecutor, graphwise_error_analyse
from ppq.api import ENABLE_CUDA_KERNEL, export_ppq_graph, load_onnx_graph
from ppq.quantization.optim import *

# 你需要自己写一个函数来加载数据
calibration_dataloader = [torch.rand(size=[1, 3, 640, 640]).cuda() for _ in range(32)]

with ENABLE_CUDA_KERNEL():
    graph = load_onnx_graph(onnx_import_file='Models/Temp/yolox_l_no_qdq.onnx')
    graph.remove_variable(graph.variables['dets'])
    graph.remove_variable(graph.variables['labels'])
    graph.remove_operation(graph.operations['TRTBatchedNMS_535'])
    graph.mark_variable_as_graph_output(graph.variables['1556'])
    graph.mark_variable_as_graph_output(graph.variables['1557'])

    quantizer   = PFL.Quantizer(platform=TargetPlatform.TRT_INT8, graph=graph) # 取得 TRT_INT8 所对应的量化器
    dispatching = PFL.Dispatcher(graph=graph).dispatch(                            # 生成调度表
        quant_types=quantizer.quant_operation_types)

    # 为算子初始化量化信息
    for op in graph.operations.values():
        quantizer.quantize_operation(
            op_name = op.name, platform = dispatching[op.name])

    # 初始化执行器
    collate_fn = lambda x: x.to('cuda')
    executor = TorchExecutor(graph=graph, device='cuda')
    executor.tracing_operation_meta(inputs=torch.zeros(size=[1, 3, 640, 640]).cuda())
    executor.load_graph(graph=graph)

    # 创建优化管线，在 ProgramEntrance_1.py 中，我们使用 QuantizationSetting 创建优化管线
    # 而在这里，所有的一切需要你手动创建
    pipeline = PFL.Pipeline([
        QuantizeSimplifyPass(),
        QuantizeFusionPass(
            activation_type=quantizer.activation_fusion_types),
        ParameterQuantizePass(),
        RuntimeCalibrationPass(),
        PassiveParameterQuantizePass(),
        QuantAlignmentPass(force_overlap=True),

        # 微调你的网路
        # LearnedStepSizePass(steps=500)

        # 如果需要训练微调网络，训练过程必须发生在 ParameterBakingPass 之前
        # ParameterBakingPass()

        # 启动保序校准，在 graphwise_error_analyse 中保序校准精度可能低于正常的 percentile 校准
        # 但这一校准方法是专为分类变量设计的，它的分类精度应当更高
        # IsotoneCalibrationPass(variables=[name for name in graph.outputs])
    ])

    # 调用管线完成量化
    pipeline.optimize(
        graph=graph, dataloader=calibration_dataloader, verbose=True, 
        calib_steps=32, collate_fn=collate_fn, executor=executor)

    graphwise_error_analyse(
        graph=graph, running_device='cuda', dataloader=calibration_dataloader, 
        collate_fn=lambda x: x.cuda())

    export_ppq_graph(
        graph=graph, platform=TargetPlatform.TRT_INT8, 
        graph_save_to='Output/quantized.onnx', 
        config_save_to='Output/quantized.json')
    
    from ppq.utils.TensorRTUtil import build_engine, Benchmark, Profiling
    build_engine(onnx_file='Output/quantized.onnx', engine_file='Output/INT8.engine', int8=True, int8_scale_file='Output/quantized.json')
    build_engine(onnx_file='Output/quantized.onnx', engine_file='Output/FP16.engine', int8=False)
    Benchmark('Output/INT8.engine')
    Benchmark('Output/FP16.engine')
    Profiling('Output/INT8.engine')

**Comment by shuyuan-wang** [comment]
用ProgramEntrance2却是复现出来了，但是用ProgramEntrance1就会出现慢的问题，2里面是更新了什么吗？

**Comment by ZhangZhiPku** [comment]
我好像只是给nms删了？

**Comment by shuyuan-wang** [comment]
感觉这个问题可以先放一放，我遇到了另一个问题，我使用的onnx都是从mmdeploy转tensorrt的时候转出来的带tensorrt plugin算子的onnx。使用yolox_s的FP16 engine作为比较，使用mmdeploy转出来的engine Average Precision是0.403，但是用ppq后，会先出来quantized.onnx，然后使用这个onnx转成fp16的engine。这个出来的AP就变成了0.395。我用的就是你前面写的那个文件。想请问一下中间转quantized.onnx的时候是只simplifier了一下吗，还是有改动？
